packages/wallet-core/src/protocols/direct-funder.ts

```diff
@@ -0,0 +1,262 @@
+import {Message} from '../wire-protocol';
+import {BN} from '../bignumber';
+import {addHash, hashState, signState} from '../state-utils';
+import {Address, SignatureEntry, State, StateWithHash, Uint256} from '../types';
+
+type AddressedMessage = {recipient: string; message: Message};
+
+export type OpenChannelEvent =
+  | {type: 'MessageReceived'; message: Message}
+  | {type: 'FundingUpdated'; amount: Uint256; finalized: boolean};
+
+type SignedStateHash = {hash: string; signatures: SignatureEntry[]};
+
+export enum WaitingFor {
+  theirPreFundSetup = 'DirectFunder.theirPreFundSetup',
+  safeToDeposit = 'DirectFunder.safeToDeposit',
+  channelFunded = 'DirectFunder.channelFunded',
+  theirPostFundState = 'DirectFunder.theirPostFundSetup'
+}
+
+export type OpenChannelObjective = {
+  status: WaitingFor | 'success' | 'error';
+  channelId: string;
+  openingState: State;
+  myIndex: number;
+
+  preFS: SignedStateHash;
+  // TODO: (ChainService) The asset class is _ignored_ here.
+  funding: {amount: Uint256; finalized: boolean};
+  // TODO: (ChainService) We will need to store funding requests once this gets hooked up to a chain service
+  fundingRequests: {tx: string}[];
+  postFS: SignedStateHash;
+};
+
+export type Action =
+  | {type: 'sendMessage'; message: AddressedMessage}
+  // TODO: (ChainService) We will need to include more data once this gets hooked up to a chain service
+  | {type: 'deposit'; amount: Uint256};
+
+export type OpenChannelResult = {
+  objective: OpenChannelObjective;
+  actions: Action[];
+};
+
+/**
+ *
+ * @param objective: A rich OpenChannelObjective data structure storing relevant data to the objective
+ * @param event an OpenChannelEvent that can trigger a state transition + actions
+ * @param myPrivateKey
+ * @returns OpenChannelResult, a data structure containing
+ *            - the current state of the objective
+ *            - actionst to be triggered by an imperative shell
+ *
+ * This is a state machine implementation of a protocol for opening a directly funded channel.
+ * It operates on a "rich" OpenChannelObjective state, which stores:
+ * 1. the channel's initial state
+ * 2. the hash of the expected preFS and its hash
+ * 3. the on-chain funding of the channel
+ * 4. funding requests
+ * 5. the hash of the expected preFS and its hash
+ *
+ * The machine receives either messages or chain events.
+ * It _whitelists_ states, rejecting any state other than the expected state.
+ *
+ * A wallet implementation can then use the result with this sequence of asynchronous operations
+ * 1. record the new objective state as well as the resulting actions
+ * 2. trigger the resulting actions asynchronously
+ * 3. mark the actions as being successful
+ *
+ * If the wallet crashes after 1 & before 3, the wallet can decide to re-trigger
+ * the actions on a case-by-case basis, based on whether the action is safe to
+ * re-trigger.
+ */
+export function openChannelCranker(
```

Comment:
This function is slightly long so it might be nice to split it into  `processEvents` /`processActions` functions?